### PR TITLE
refactor(undo.c): resolve clint issues

### PR DESCRIPTION
**Pull request by lewis6991**

### CODE REVIEW

**Comment by zeertzjq** [comment]
It seems bufhl test is expecting an intermediate screen state. Changing `change3` to `changes` passes without this PR.

File: src/nvim/undo.c [hunk 1]
```diff
@@ -1622,17 +1572,17 @@ void u_read_undo(char *name, const char_u *hash, const char_u *orig_name FUNC_AT
     }
     if (old_header_seq > 0 && old_idx < 0 && uhp->uh_seq == old_header_seq) {
       assert(i <= SHRT_MAX);
```
Comment (zeertzjq) [inline src/nvim/undo.c:1574]: I think these assertions need change.

File: src/nvim/undo.c [hunk 2]
```diff
@@ -1980,6 +1909,14 @@ void undo_time(long step, bool sec, bool file, bool absolute)
 
   // "target" is the node below which we want to be.
   // Init "closest" to a value we can't reach.
```
Comment (zeertzjq) [inline src/nvim/undo.c:1911]: Comment to far away from what it describes

File: src/nvim/undo.c (hunk 1 above)
Comment (lewis6991) [reply src/nvim/undo.c:1574]: Why so?

Comment (zeertzjq) [reply src/nvim/undo.c:1574]: `INT16_MAX`

File: src/nvim/undo.c [hunk 3]
```diff
@@ -1621,18 +1571,18 @@ void u_read_undo(char *name, const char_u *hash, const char_u *orig_name FUNC_AT
       }
     }
     if (old_header_seq > 0 && old_idx < 0 && uhp->uh_seq == old_header_seq) {
-      assert(i <= SHRT_MAX);
-      old_idx = (short)i;
+      assert(i <= UINT16_MAX);
```
Comment (zeertzjq) [inline src/nvim/undo.c:1574]: It's `INT16_MAX`, not `UINT16_MAX`

**Comment by zeertzjq** [comment]
There is a commit I found: <https://github.com/neovim/neovim/commit/c0fd830be4ef3eafb756c463aee6408323ba4e58>

File: src/nvim/undo.c [hunk 4]
```diff
@@ -2041,24 +1978,24 @@ void undo_time(long step, bool sec, bool file, bool absolute)
...
     }
   }
-  closest_start = closest;
-  closest_seq = curbuf->b_u_seq_cur;
+  long closest_start = closest;
+  long closest_seq = curbuf->b_u_seq_cur;
 
   // When "target" is 0; Back to origin.
+  int mark;
+  int nomark = 0;  // shut up compiler
```
Comment (zeertzjq) [inline src/nvim/undo.c:1986]: Better move these two up two lines

**Review by zeertzjq — APPROVED** [review]

**Pull request merged**